File: LeetCode/Problems/Python/653. Two Sum IV - Input is a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: TreeNode, k: int) -> bool:
        if not root:
            return False
        
        
        queue = [root]
        lookUp = set()
        
        while queue:
            node = queue.pop()
            if not node:
                continue
            if k-node.val in lookUp:
                return True
            lookUp.add(node.val)
            queue.append(node.left)
            queue.append(node.right)
        return False

File: LeetCode/Problems/C++/653. Two Sum IV - Input is a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//dfs recursion
//Time: O(n)
//Space: O(n)
class Solution {
public:
    bool dfs(TreeNode* node, unordered_set<int>& lookup, int k){
        if(node == NULL) return false;
        if(lookup.count(k-node->val)){
            return true;
        } 
        
        lookup.insert(node->val);
        return dfs(node->left, lookup, k) || dfs(node->right, lookup, k);  
    }
    
    bool findTarget(TreeNode* root, int k) {
        unordered_set<int> lookup;
        
        return dfs(root, lookup, k);

    }
};